examples/nodejs/storage/user-program.ts

```diff
@@ -0,0 +1,98 @@
+      const value = getResponse.value;
+
+      // eg can do `const value: string = getResponse.value as string;`
+      console.log(`The value is: ${value} and the type is ${getResponse.type}`);
+
+      // To request back as a concrete type means the developer knows the type.
+      // If the developer is right, they get the value. And if they are wrong,
+      // we have two choices:
+      // 1. return undefined, or
+      // 2. throw an exception.
+      //
+      // Returning undefined is a silent failure. The developer may not know
+      // that the value was not what they expected. They may also confuse
+      // the undefind with the value not in the store.
+      //
+      // Throwing an exception is a loud failure. The developer will know
+      // immediately they were wrong. But not all developers want to
+      // handle exceptions.
+      //
+      // We can make it easy for the developer to exhaustively check the type
+      // and handle the cases they care about as follows:
+      switch (getResponse.type) {
+        case StoreValueType.STRING:
+          console.log('The value is a string:', getResponse.tryGetValueString());
+          break;
+        case StoreValueType.INTEGER:
+          console.log('The value is an integer:', getResponse.tryGetValueInteger());
+          break;
+        case StoreValueType.FLOAT:
+          console.log('The value is a float:', getResponse.tryGetValueFloat());
+          break;
+        case StoreValueType.BOOLEAN:
+          console.log('The value is a boolean:', getResponse.tryGetValueBoolean());
+          break;
+        default:
+          console.error('Unknown type');
+      }
+      // An alternative to this switch is:
+      // - StoreGet.Success has a `value` field which is of type `StoreValue`
+      // `StoreValue` is a base class with derived types, each with a `value` field.
+      // - StoreStringValue with a string-valued `value` field,
+      // - StoreIntegerValue with an integer-valued `value` field,
+      // - StoreFloatValue, etc.
+      // In that world the developer can use more instanceof checks:
+      // if (getResponse instanceof StoreGet.Success) {
+      //    const storeValue = getResponse.value;
+      //    if (storeValue instanceof StoreStringValue) {
+      //      console.log('The value is a string:', storeValue.value);
+      //    } else if (storeValue instanceof StoreIntegerValue) {
+      //      console.log('The value is an integer:', storeValue.value);
+      //    } else if (storeValue instanceof StoreFloatValue) {
+      //      // ...
+      //    } else {
+      //      console.error('Unknown type');
+      //    }
+      // }
+      //
+      // I feel this pattern makes sense in a language with first-class pattern matching, such as
+      // Rust, Scala, Kotlin, etc.
+      // But in JavaScript, I feel the switch is at the right level of complexity.
```

Comment:
I agree that the switch seems like it would be nicer in typescript than the instanceof. It'd be really awesome though to find a way to be able to use a switch and still get the strong typing inside of the case branches. This SO post looks really interesting:

https://stackoverflow.com/a/56597745

I wonder if we could do that and sort of get the best of both worlds?

Follow up question will be: if we decide to go down one of these paths that differs from our `instanceof` pattern in the CacheClient, is it going to feel weird that the persistent store client has a very different API? If so, is there any forward-compatible way for us to make the CacheClient API feel more similar to whatever we decide to do here? Or, maybe the persistent store API lands in a 2.0 version of the SDK where the CacheClient one changes to match it?

Reply:
@bruuuuuuuce has a compelling PR open [here](https://github.com/momentohq/client-sdk-javascript/pull/1254) where explored there. There are TS options that can enforce exhaustive enum coverage on a switch.